File: Lowest_Common_Ancestor/LCA.py
class Node:
    def __init__(self, info):
        self.info = info
        self.right = None
        self.left = None
        self.level = None

    def __str__(self):
        return str(self.info)


class BST:
    def __init__(self):
        self.root = None

    def build(self, val):
        if self.root == None:
            self.root = Node(val)
        else:
            temp = self.root
            while True:
                if val < temp.info:
                    if temp.left:
                        temp = temp.left
                    else:
                        temp.left = Node(val)
                        break
                elif val > temp.info:
                    if temp.right:
                        temp = temp.right
                    else:
                        temp.right = Node(val)
                        break
                else:
                    break

def lca(root, search1, search2):
    if root is None:
        return root
    if ((search1 < root.info) and (search2 < root.info)):
        root = lca(root.left,search1,search2)
    elif ((search1 > root.info) and (search2 > root.info)):
        root = lca(root.right,search1,search2)
    elif ((search1 < root.info) and (search2 > root.info)):
        return root
    return root

class DAGNode:
    def __init__(self, info):
        self.info = info
        self.suc = []
        self.pre = []


def dag_lca(root, search1, search2):
    if type(root) != DAGNode or type(search1) != DAGNode or type(search2) != DAGNode:
        return None
    if root is None:
        return None
    if root == search1 or root == search2:
        return root
    if search1 == search2:
        return search1.info

    result = []
    for i in range(len(search1.pre)):
        for j in range(len(search2.pre)):
            if (search1.pre[i].info == search2.pre[j].info):
                result.append(search1.pre[i].info)

    if (result == []):
        if (search1.info > search2.info):
            result.append(dag_lca(root, search1.pre[0], search2))
        else:
            result.append(dag_lca(root, search1, search2.pre[0]))

    return max(result)


#for building tree
#tree = BST()
#tree.build(2)
#tree.build(1)
#tree.build(3)

#intial build using user input
# print("Enter the number of nodes")
# size= int(input())

# takes node values
#print("Enter Values")
#for i in range(size):
    #values = int(input())
    #tree.build(values)


#print("First node for lca:")
#search1 = int(input())
#print("Second node for lca:")
#search2 = int(input())
#result = lca(tree.root, search1, search2)
#print(result.info)

File: Lowest_Common_Ancestor/lca.java
//import java.util.Scanner;

class lca {
    class Node {
        Node left;
        Node right;
        int data;

        Node(int data) {
            this.data = data;
            left = null;
            right = null;
        }
    }

    Node root;

    // intialise root
    lca() {
        root = null;
    }

    void insert(int data) {
        root = insert_RecursiveNode(root, data);
    }

    Node insert_RecursiveNode(Node root, int data) {
        // tree empty
        if (root == null) {
            root = new Node(data);
            return root;
            // traverse tree
        } else {
            Node temp;
            if (data <= root.data) {
                temp = insert_RecursiveNode(root.left, data);
                root.left = temp;
            } else {
                temp = insert_RecursiveNode(root.right, data);
                root.right = temp;
            }
            return root;
        }
    }

    public static Node lcaSol(Node root, int search1, int search2) {
        if (root == null) {
            return null;
        } else if (root.data == search1 || root.data == search2) {
            return root;
        }
        // lca is on left subtree
        else if (root.data > search1 && root.data > search2) {
            return lcaSol(root.left, search1, search2);
        }
        // lca is on right
        else if (root.data < search1 && root.data < search2) {
            return lcaSol(root.right, search1, search2);
        }
        return root;
    }

    // as per the constrains no value shall be repeated in the tree and nodes need
    // to be unique.
    // public static void main(String[] args) {
    // was done using user input where tree wasn't intialised.
    // Scanner input = new Scanner(System.in);
    // System.out.print("Please enter number of nodes first, then followed by
    // values.");
    // int numNodes = input.nextInt();
    // Node root = null;
    // while (numNodes-- > 0) {
    // int value = input.nextInt();
    // root = insert(root, value);
    // // System.out.print(root.data);
    // }
    // System.out.print("Enter the two value for which you want to find LCA:");
    // int search1 = input.nextInt();
    // int search2 = input.nextInt();
    // input.close();
    // Node lca = lca(root, search1, search2);
    // System.out.print(lca.data);

    // intialise tree
    // lca bst = new lca();
    // System.out.print("Got intialised");
    // bst.insert(45);
    // bst.insert(10);
    // bst.insert(7);
    // bst.insert(12);
    // bst.insert(90);
    // bst.insert(50);
    // Node resultNode = lcaSol(bst.root, 7, 90);
    // System.out.print(resultNode.data);
    // }
}
